### PR TITLE
[ml] raise error when serializing Predictor

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

Predictors are generally not serializable. This can also lead to undesirable situations that are often not immediately obvious such as when instantiating a Predictor locally when intending to do so remotely.

This change raises an error whenever a Predictor instance is serialized so that the user gets immediate feedback.

## Related issue number

Example: https://github.com/ray-project/ray/pull/23245#discussion_r828435934

## Checks

Added a simple unit test.

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
@Yard1 does this mean we can change TFPredictor to accept the model instance instead of a model creator function, if we are no longer expecting predictors to be serializable?

File: python/ray/ml/predictor.py [hunk 1]
```diff
@@ -38,3 +44,8 @@ def predict(self, data: DataBatchType, **kwargs) -> DataBatchType:
             DataBatchType: Prediction result.
         """
         raise NotImplementedError
+
+    def __reduce__(self):
+        raise PredictorNotSerializableException(
+            "Predictor instances are not serializable."
```
Comment (ericl) [inline python/ray/ml/predictor.py:50]: Tell the user they can pass checkpoint instead?

**Review by ericl — APPROVED** [review]
Probably will need to fix other tests after this.

**Comment by Yard1** [comment]
@amogkam I can make it accept both

**Pull request merged**